### PR TITLE
IBM Cloud deployment improvements

### LINKED ISSUE #423: Deployment: update infrastructure to be deployed in IBM cloud
### What is the expected enhancement?

For the `dev-forum` we want to be able to prepare different demos and have everything ready for users to be able to launch the project. We have different PR opened for this use-case:

1. #399 : where we are trying to finish with the configuration of docker-compose to run it locally.
2. ~~#255 : to be able to launch `quantum-serverless` in IBM Cloud. This PR is old and it didn't have the `gateway` so probably a lot of changes will be needed but we can take it as reference.~~
3. #447 : previous work will continue in this PR.

For `helm` a configuration that we are going to be able to drop will be the sidecars probably.

### PR DESCRIPTION
**Pull request by Tansito**
### Summary

Testing our current configuration in IBM Cloud. Fix #423 

### Details and comments

Due to the differences between the deployment in local vs IBM Cloud I made more dynamic our Ingress configuration with the purpose to be able to support AWS too without the need to change the template if not only the values (as a normal user would do).

- I removed `nginxconfig` in favor of adding those annotations in `ingress`.
- Renamed `ray-ingress` to just `ingress` now that we are going to expose different applications
- `ingress` now is more dynamic so you can add the annotations and host configuration through the values file.
- How only the `gateway` + `repository` are going to be exposed I changed the `serviceType` of almost all the services to `ClusterIP`.
- Fix in terraform values that now are required by the project: `machine_type`, `workers_per_zone`...
- Removed `helm` integration in `terraform` due to now helm release requires values from the cluster after its creation (investigating if we can obtain them).
- `ray-cluster` changed from a fixed worker to a dynamic set of workers

### Resources

- [IBM Cloud](https://cloud.ibm.com/docs/containers?topic=containers-managed-ingress-setup#managed-ingress-steps-resource) 
- [NGINX Controller annotations](https://kubernetes.github.io/ingress-nginx/user-guide/nginx-configuration/annotations/#rewrite)

### CODE REVIEW

File: infrastructure/helm/quantumserverless/charts/gateway/templates/deployment.yaml [hunk 1]
```diff
@@ -33,19 +33,19 @@ spec:
             {{- toYaml .Values.securityContext | nindent 12 }}
           image: "{{ .Values.image.repository }}:{{ .Values.image.tag | default .Chart.AppVersion }}"
           imagePullPolicy: {{ .Values.image.pullPolicy }}
-          command: ["python", "manage.py", "runserver", "0.0.0.0:8000"] #[ "gunicorn", "gateway.wsgi:application", "--bind", "0.0.0.0:8000", "--workers=3" ]
+          command: [ "gunicorn", "main.wsgi:application", "--bind", "0.0.0.0:8000", "--workers=4" ]
```
Comment (Tansito) [inline infrastructure/helm/quantumserverless/charts/gateway/templates/deployment.yaml:36]: Right now we the port in unicorn is `fixed`. I need to make it dynamic.

File: infrastructure/helm/quantumserverless/charts/repository/templates/deployment.yaml [hunk 2]
```diff
@@ -33,6 +33,7 @@ spec:
             {{- toYaml .Values.securityContext | nindent 12 }}
           image: "{{ .Values.image.repository }}:{{ .Values.image.tag | default .Chart.AppVersion }}"
           imagePullPolicy: {{ .Values.image.pullPolicy }}
+          command: [ "gunicorn", "main.wsgi:application", "--bind", "0.0.0.0:8060", "--workers=4" ]
```
Comment (Tansito) [inline infrastructure/helm/quantumserverless/charts/repository/templates/deployment.yaml:36]: Same as `gateway`, I need to make this port _dynamic_.

File: infrastructure/helm/quantumserverless/values-ibm.yaml [hunk 3]
```diff
@@ -0,0 +1,318 @@
+# ===================
```
Comment (Tansito) [inline infrastructure/helm/quantumserverless/values-ibm.yaml:1]: How the `values` are slightly different from `local` and a 3rd party `provider` I created a separated value. We can discuss it.

File: infrastructure/helm/quantumserverless/values.yaml [hunk 4]
```diff
@@ -97,43 +85,23 @@ rayClusterEnable: true
...
-          - --email-domain="*"
-          - --insecure-oidc-allow-unverified-email=true
-          - --http-address=0.0.0.0:4180
-          - --cookie-secret=SECRET0123456789
-          - --provider=keycloak-oidc
-          - --client-id=rayclient
-          - --upstream="http://HELM-RELEASE-kuberay-head-svc:8265"
-          - --redirect-url=http://localhost/oauth2/callback
+      dashboard-host: "0.0.0.0"
+    ports: [{containerPort: 8265, name: dashboard}]
```
Comment (Tansito) [inline infrastructure/helm/quantumserverless/values.yaml:95]: How the `gateway` will be the `entrypoint` for the project I started to remove the `sidecarContainers` from other services.

File: infrastructure/helm/quantumserverless/values.yaml [hunk 5]
```diff
@@ -316,9 +263,9 @@ kube-prometheus-stack:
         allow_sign_up: "true"
         client_id: grafana
         client_secret: GRAFANASECRET-CHANGEME
-        api_url: "http://LOCAL-IP:31059/realms/quantumserverless/protocol/openid-connect/userinfo"
-        auth_url: "http://LOCAL-IP:31059/realms/quantumserverless/protocol/openid-connect/auth"
-        token_url: "http://LOCAL-IP:31059/realms/quantumserverless/protocol/openid-connect/token"
+        api_url: "http://RELEASE_NAME-keycloak:31059/realms/quantumserverless/protocol/openid-connect/userinfo"
```
Comment (Tansito) [inline infrastructure/helm/quantumserverless/values.yaml:266]: I wasn't sure about this @akihikokuroda @psschwei so you can correct me if I'm wrong but I think we can use the service internally? Let me know what you think.

**Comment by Tansito** [comment]
@psschwei I don't know if you can help me debugging a problem that I found. Installing the helm I'm not being able to find `kuberay/operator`:

Screenshot 2023-04-18 at 12 12 39

**Comment by psschwei** [comment]
> @psschwei I don't know if you can help me debugging a problem that I found. Installing the helm I'm not being able to find `kuberay/operator`:

Looks like the tag should be `v0.5.0`

**Comment by Tansito** [comment]
Totally right @psschwei , hours fighting yesterday and there was a `v` 🤦 , thank you!

File: infrastructure/helm/quantumserverless/values.yaml (hunk 5 above)
Comment (akihikokuroda) [reply infrastructure/helm/quantumserverless/values.yaml:270]: I think that I tried but it didn't work. The uses`http://RELEASE_NAME-keycloak:31059` as the base address for the redirect.  It must be an address that the browser can access to.

File: infrastructure/helm/quantumserverless/values.yaml (hunk 4 above)
Comment (akihikokuroda) [reply infrastructure/helm/quantumserverless/values.yaml]: The port **8265** is used by the Ray console as well as the APIs accessed through the gateway.  This side car probably still is needed.

**Comment by psschwei** [comment]
Is this PR the same config we used to set up the cluster for the dev forum?

**Comment by Tansito** [comment]
Yes but with some on-real-time changes 😂 . Like for example we improved the `raycluster` configuration, we needed to modify the `realm` configuration due to some kind of `random` error and a problem with the docker `entrypoints` that we need to take a look before open it to review. But yeah is almost with these changes.

**Comment by Tansito** [comment]
I'm working on those points right now btw.

**Comment by Tansito** [comment]
Ok, there is two points that it remains to solve before open the PR:
- Solve the problem with the Docker **ENTRYPOINTS**: they are not being executed when applying the helm (@psschwei I don't know if you suffered it before).
- Add users to the keycloak-realm: @IceKhan13 @akihikokuroda , how did you export users in the past?

**Comment by pacomf** [comment]
@Tansito tests are failing because: https://github.com/Qiskit-Extensions/quantum-serverless/pull/509#issuecomment-1538669209

**Comment by Tansito** [comment]
Yes, I was commenting w/ @psschwei and @IceKhan13 in the issue, thank you @pacomf !

**Comment by IceKhan13** [comment]
Any of #508 and #512 will fix failing test :)

**Comment by IceKhan13** [comment]
@Tansito can you merge main here? tests are fixed now :)

File: infrastructure/helm/quantumserverless/charts/gateway/templates/deployment.yaml [hunk 6]
```diff
@@ -33,19 +33,19 @@ spec:
             {{- toYaml .Values.securityContext | nindent 12 }}
           image: "{{ .Values.image.repository }}:{{ .Values.image.tag | default .Chart.AppVersion }}"
           imagePullPolicy: {{ .Values.image.pullPolicy }}
-          command: [ "/usr/src/app/entrypoint.sh", "gunicorn", "main.wsgi:application", "--bind", "0.0.0.0:8000", "--workers=3" ]
+          args: [ "gunicorn", "main.wsgi:application", "--bind", "0.0.0.0:{{ .Values.service.port }}", "--workers=4" ]
```
Comment (IceKhan13) [inline infrastructure/helm/quantumserverless/charts/gateway/templates/deployment.yaml:36]: ah, that was a problem of not having migrations :)

File: infrastructure/helm/quantumserverless/charts/gateway/templates/deployment.yaml [hunk 7]
```diff
@@ -33,19 +33,19 @@ spec:
...
               protocol: TCP
-          livenessProbe:
-            httpGet:
-              path: /metrics
-              port: http
-          readinessProbe:
-            httpGet:
-              path: /metrics
-              port: http
+#          livenessProbe:
```
Comment (IceKhan13) [inline infrastructure/helm/quantumserverless/charts/gateway/templates/deployment.yaml:41]: My only comment is this. Maybe we should uncomment this? :) 
or `/metrics` is not available? or do `/api/v1/`

**Comment by psschwei** [comment]
this failed to install for me:
```
$ helm -n quantum-serverless install quantum-serverless --create-namespace -f values-ibm.yaml .
coalesce.go:223: warning: destination for loki.gateway.affinity is a table. Ignoring non-table value (podAntiAffinity:
  requiredDuringSchedulingIgnoredDuringExecution:
    - labelSelector:
        matchLabels:
          {{- include "loki.gatewaySelectorLabels" . | nindent 10 }}
      topologyKey: kubernetes.io/hostname
)
Error: INSTALLATION FAILED: unable to build kubernetes objects from release manifest: [resource mapping not found for name: "loki" namespace: "" from "": no matches for kind "GrafanaAgent" in version "monitoring.grafana.com/v1alpha1"
ensure CRDs are installed first, resource mapping not found for name: "loki" namespace: "" from "": no matches for kind "LogsInstance" in version "monitoring.grafana.com/v1alpha1"
ensure CRDs are installed first, resource mapping not found for name: "loki" namespace: "" from "": no matches for kind "PodLogs" in version "monitoring.grafana.com/v1alpha1"
ensure CRDs are installed first]

```

(looks like I broke something in https://github.com/Qiskit-Extensions/quantum-serverless/commit/ce747be61d362708ebf446821aec1659cb20911b -- seems like we need the grafana-agent CRDs or maybe even the entire operator...)

not for this PR, but we may want to look into something like kustomize for some of the provider-specific overlays: https://jfrog.com/blog/power-up-helm-charts-using-kustomize-to-manage-kubernetes-deployments/

also, one minor nit: can we update the readme for IBM-specific install instructions?

File: infrastructure/helm/quantumserverless/charts/gateway/templates/deployment.yaml (hunk 7 above)
Comment (Tansito) [reply infrastructure/helm/quantumserverless/charts/gateway/templates/deployment.yaml:41]: You are totally right. My two cents here from previous experiences is that the security team will ask us for **two different** end-points for `liveness` and `readiness` for a simple HTTP 200 and a DB access. For the DB access I think `/api/v1` can work (I would need to confirm it) but for the HTTP 200 we will need a specific end-point. That's why I commented them. If you are agree @IceKhan13 I can create an issue for this.

**Comment by Tansito** [comment]
> (looks like I broke something in https://github.com/Qiskit-Extensions/quantum-serverless/commit/ce747be61d362708ebf446821aec1659cb20911b -- seems like we need the grafana-agent CRDs or maybe even the entire operator...)

@psschwei I'm going to update this PR with `master` now but my last deployment in IBM Cloud was with your changes and I think it was working. I can confirm you in some minutes when I try to deploy again.

> we may want to look into something like kustomize for some of the provider-specific overlays

I'm always open for improvements. Before introduce more things in the infrastructure I would like to work too in the tests: #117 and #118 

> also, one minor nit: can we update the readme for IBM-specific install instructions?

You are totally right. I will review it now.

File: infrastructure/helm/quantumserverless/charts/gateway/templates/deployment.yaml (hunk 6 above)
Comment (Tansito) [reply infrastructure/helm/quantumserverless/charts/gateway/templates/deployment.yaml:36]: mostly, it seems that `command` overrides the `entrypoint` that you can have in your container (you always learn something new).

**Comment by Tansito** [comment]
@psschwei I can confirm you. Regardless the warning in:

```
coalesce.go:223: warning: destination for loki.gateway.affinity is a table. Ignoring non-table value (podAntiAffinity:
  requiredDuringSchedulingIgnoredDuringExecution:
    - labelSelector:
        matchLabels:
          {{- include "loki.gatewaySelectorLabels" . | nindent 10 }}
      topologyKey: kubernetes.io/hostname
)
```

I could deploy it in IBM Cloud without a problem.

**Comment by psschwei** [comment]
Yeah, the warning has been there for a while (I think I opened an issue for it sometime back)... the errors were new for me (and also weird, because I didn't get them before...). 

On kustomize I'm thinking two things: 
* minimizing duplicates in the config files 
* making it easy for users to set their values (secrets, etc, basically all those things in the readme we say for them to edit)

not any hurry on that, more of a day 2 / nice to have at this point. Agree testing would be better for now :smile:

**Comment by psschwei** [comment]
> kustomize

Someone this week recommended [helmfile](https://github.com/helmfile/helmfile) as a potentially better fit, so we have options :smile:

**Review by IceKhan13 — APPROVED** [review]
monumental work of setting this all up 👏

**Review by pacomf — APPROVED** [review]

**Review by pacomf — APPROVED** [review]
Here we go! 🚀

**Pull request merged**